presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java

```diff
@@ -561,6 +566,120 @@ private void invokeNoMoreSplitsIfNecessary()
+            InputSplit[] targetSplits = hdfsEnvironment.doAs(
+                    hdfsContext.getIdentity().getUser(),
+                    () -> fileInputFormat.getSplits(targetJob, 0));
+
+            InternalHiveSplitFactory splitFactory = new InternalHiveSplitFactory(
+                    targetFilesystem,
+                    partitionName,
+                    inputFormat,
+                    schema,
+                    partitionKeys,
+                    effectivePredicate,
+                    partitionMatchSupplier,
+                    partition.getTableToPartitionMapping(),
+                    Optional.empty(),
+                    getMaxInitialSplitSize(session),
+                    isForceLocalScheduling(session),
+                    s3SelectPushdownEnabled);
+            lastResult = addSplitsToSource(targetSplits, splitFactory);
+            if (stopped) {
+                return COMPLETED_FUTURE;
+            }
+        }
+        return lastResult;
+    }
+
+    private Iterator<InternalHiveSplit> buildManifestFileIterator(Table table,
+            Path parent,
+            List<Path> paths,
+            DirectoryLister directoryLister,
+            NamenodeStats namenodeStats,
+            InternalHiveSplitFactory splitFactory,
+            boolean splittable)
+            throws DirectoryNotAllowedException
+    {
+        Map<Path, LocatedFileStatus> fileStatuses = new HashMap();
+        try {
+            FileSystem targetFilesystem = hdfsEnvironment.getFileSystem(hdfsContext, parent);
+            HiveFileIterator fileStatusIterator = new HiveFileIterator(table, parent, targetFilesystem, directoryLister, namenodeStats, IGNORED, true);
+            fileStatusIterator.forEachRemaining(status -> fileStatuses.putIfAbsent(status.getPath(), status));
+        }
+
+        catch (IOException e) {
+            throw new PrestoException(HIVE_FILE_NOT_FOUND, String.format("No Filesystem available for %s", parent), e);
+        }
+
+        List<LocatedFileStatus> locatedFileStatuses = paths.stream().map(this::canonicalizeSymlinkPath).map(path -> fileStatuses.get(path)).collect(Collectors.toList());
+        if (locatedFileStatuses.contains(null)) {
+            // This check will catch all directories in the manifest since HiveFileIterator will not return any directories
+            throw new DirectoryNotAllowedException();
+        }
+
+        return hdfsEnvironment.doAs(hdfsContext.getIdentity().getUser(), () -> locatedFileStatuses.stream()
+                .map(locatedFileStatus -> splitFactory.createInternalHiveSplit(locatedFileStatus, OptionalInt.empty(), splittable, Optional.empty()))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .iterator());
+    }
+
+    public static class DirectoryNotAllowedException
+            extends Exception
+    {
+    }
+
+    private Path canonicalizeSymlinkPath(Path path)
```

Comment:
it seems canonicalization is implementation specific, e.g `org.apache.hadoop.hdfs.DistributedFileSystem#canonicalizeUri`, but `FileSystem#canonicalizeUri` is not exposed.

I think we can use `org.apache.hadoop.fs.Path#getPathWithoutSchemeAndAuthority` for normalization as all `targetPaths` share same `parent`

Reply:
Good point. This is a holdover from supporting multiple parents. I will update.